### PR TITLE
feat!: rewrite external copy with multi version hoisting support

### LINKED ISSUE #572: Issues with flatten extrernals tracing and multiple versions
On production builds, we trace used package files from `node_modules` and extract them to `.output/server/node_modules` but at the same time, flatten the structure by choosing the latest requested version of each package.

This strategy has two major issues:
 1. In the case of two major versions (sub-dependencies)
 2. In the case of two minor versions (eg #259) since we already traced, if the version is not compatible one will fail as we already traced paths back to their actual package path

For solving 1, only possible solution is hoisting sub-dependencies into `node_modules/parent/node_modules/child`. It can be also done by **respecting package manager structure without flattening it** or less fragile by doing what package managers do for hosting with more stable output with different pms!

For solving 2, we need to **re-trace internal package paths** in case of a conflict is detected as **external exports and API** of a package should be compatible in semver-minor/patch but since we already traced them to internal structure, it will be broken.

### PR DESCRIPTION
**Pull request by pi0**
Resolves #572

Based on amazing work from @mahdiboomeri (#758) this PR introduces a full rewrite of the external copy mechanism.

Previously if multiple versions of an external library were being used by traced externals, we used this strategy:
- In case of minor version conflicts, pick the latest to use for all libs
- In case of major version conflicts, warn and pick the latest semver-major one

This causes unresolvable issues especially when sub-dependencies are conflicting.

With this refactor, we respect hoisted structure of package manager and map it into `.output/server/node_modules`:
- Create a normalized list of all traced files and their parents
- Create a normalized list of all relevant package/versions from traced files
- For each package
  - If there is only one traced version detected, copy it as as
  - If more than one traced version is detected, copy to `node_modules/{name}@{version}` and symlink to `node_modules/{parent}/node_modules/{name}` for each parent. If no parent modules are detected, hoist it to top level as default (user's version)

In case of having issues with new implementation in nitropack v2, you can opt to use legacy externals:

```js
experimental: {
  // Use nitropack v1 behavior
  legacyExternals: true
}
```

### CODE REVIEW

**Comment by pi0** [comment]
@mahdiboomeri If you check latest refactors, do you think we would still need an option for optimizeExternals?

**Comment by mahdiboomeri** [comment]
@pi0 I added the options as a workaround if they were some edge cases that weren't covered (such as a package that doesn't follow semver), but IMO the new approach seems reliable enough to skip that.

File: src/rollup/plugins/externals.ts [hunk 1]
```diff
@@ -197,102 +188,187 @@ export function externals(opts: NodeExternalsOptions): Plugin {
...
+                await writePackage(tracedPackage.name, version);
+              } else {
+                // Write alternative version into node_modules/{name}@{version}
+                await writePackage(
+                  tracedPackage.name,
+                  version,
+                  `${tracedPackage.name}@${version}`
+                );
+                // For each parent, link into node_modules/{parent}/node_modules/{name}
+                for (const parentPath of parentPkgs) {
```
Comment (mahdiboomeri) [inline src/rollup/plugins/externals.ts:345]: Linking breaks if the nested packages also have conflicting children.
Consider the following structure:

```
└── node_modules
    ├── nitro-dep-a
    │   └── nitro-lib#v1.0.0
    │       └── nested-lib#v1.0.0
    ├── nitro-dep-b
    │   └── nitro-lib#v2.0.0
    │       └── nested-lib#v2.0.0
    └── nitro-lib#v2.9.0 
```

**Comment by itpropro** [comment]
Any updates for this @pi0 ?

File: src/rollup/plugins/externals.ts (hunk 1 above)
Comment (pi0) [reply src/rollup/plugins/externals.ts:361]: Should be fixed in latest version using parent-level hoisting

**Comment by pi0** [comment]
Updated implementation with parent-level hosting for nested support, windows support using junctions (not portable), and alias support.

Although new implementation should be much better for handling hoisting situations and allow better enhancements using new clearer data-structures, it can be buggy without testing on real-world installs. So let's get it a spin. 

Also there is new `experimental.legacyExternals` flag to allows an escape hatch to use v1 behavior.

**Pull request merged**